nRF Cloud Multi Service Sample (#11900)
nrfconnect/sdk-nrf: samples/cellular/nrf_cloud_multi_service/README.rst

.. include:: /includes/external_flash_nrf91.txt

> When using CoAP, modem firmware version 1.3.5 or above is recommended, in order to benefit from the power savings provided by DTLS Connection ID.

Reviewer: Or even "... use modem firmware version..."
As we tend to avoid using "we recommend".

   Modem responses longer than this length will be replaced with an error code message (-NRF_E2BIG).
   Cannot be less than 40 bytes.

When using CoAP, the following additional configuration options are available:

.. _CONFIG_COAP_SHADOW_CHECK_RATE_SECONDS:

CONFIG_COAP_SHADOW_CHECK_RATE_SECONDS - How many seconds between requests for any change (delta) in the device shadow.

.. _CONFIG_COAP_SHADOW_THREAD_STACK_SIZE:

> CONFIG_COAP_SHADOW_THREAD_STACK_SIZE - Size in bytes of the shadow check thread stack.

Author: Thanks.

Building and running
********************

.. |sample path| replace:: :file:`samples/cellular/nrf_cloud_multi_service`

.. include:: /includes/build_and_run_ns.txt

.. _nrf_cloud_multi_service_coap_provisioning:

Provisioning a CoAP device on nRF Cloud
***************************************

For more information, see `nRF Cloud Utilities documentation`_.

Only do this step once.

.. code-block:: console

   python3 create_ca_cert.py -c US -st OR -l Portland -o "My Company" -ou "RD" -cn example.com -e admin@example.com -f CA
   Creating self-signed CA certificate...
   File created: CA0x3bc7f3b014a8ad492999c594f08bbc2fcffc5fd1_ca.pem
   File created: CA0x3bc7f3b014a8ad492999c594f08bbc2fcffc5fd1_prv.pem
   File created: CA0x3bc7f3b014a8ad492999c594f08bbc2fcffc5fd1_pub.pem

Do the following step for each device.
Execute in the same folder as the step above.

.. note::
   The options used below assume that the sample is built with the default :kconfig:option:`CONFIG_NRF_CLOUD_CLIENT_ID_SRC_IMEI` option enabled and :kconfig:option:`CONFIG_NRF_CLOUD_CLIENT_ID_PREFIX` option set to ``nrf-``.
   See :ref:`configuration_device_id` to use other device ID formats.

> .. code-block:: console
> 
>    python3 device_credentials_installer.py --ca CA0x3bc7f3b014a8ad492999c594f08bbc2fcffc5fd1_ca.pem \
>    --ca_key CA0x3bc7f3b014a8ad492999c594f08bbc2fcffc5fd1_prv.pem --id_str nrf- --id_imei -s -d --coap
>    ...
>    <- OK
>    Saving provisioning endpoint CSV file provision.csv...
>    Provisioning CSV file saved

Reviewer 2: ```suggestion
.. code-block:: console

   python3 device_credentials_installer.py --ca <certificate_serial>_ca.pem \
   --ca_key <certificate_serial>_prv.pem --id_str nrf- --id_imei -s -d --coap
   ...
   <- OK
   Saving provisioning endpoint CSV file provision.csv...
   Provisioning CSV file saved

Where ``<certificate_serial>`` is your certificate's serial number in hex.
```
The console output for the command isn't really necessary since we're only interested in the command.

.. include:: /includes/build_and_run_ns.txt

.. _nrf_cloud_multi_service_coap_provisioning:

Provisioning a CoAP device on nRF Cloud
***************************************

For more information, see `nRF Cloud Utilities documentation`_.

Only do the first step below once.

  .. code-block:: console

     python3 create_ca_cert.py -c US -st OR -l Portland -o "My Company" -ou "RD" -cn example.com -e admin@example.com -f CA
     Creating self-signed CA certificate...
     File created: CA0x3bc7f3b014a8ad492999c594f08bbc2fcffc5fd1_ca.pem
     File created: CA0x3bc7f3b014a8ad492999c594f08bbc2fcffc5fd1_prv.pem
     File created: CA0x3bc7f3b014a8ad492999c594f08bbc2fcffc5fd1_pub.pem

     python3 device_credentials_installer.py --ca CA0x3bc7f3b014a8ad492999c594f08bbc2fcffc5fd1_ca.pem \
     --ca_key CA0x3bc7f3b014a8ad492999c594f08bbc2fcffc5fd1_prv.pem --id_str nrf- --id_imei -s -d --coap
     ...
     <- OK
     Saving provisioning endpoint CSV file provision.csv...
     Provisioning CSV file saved

Once the script creates the provision.csv file, visit http://nrfcloud.com/#/provision-devices to provision the device in nRF Cloud.

> This registers the device with nRF Cloud, so that the device can use any nRF Cloud services via CoAP, MQTT, or REST connectivity.

Reviewer: via -> through
as we try to avoid Latinisms.

.. _nrf_cloud_multi_service_building_coap:

Building with CoAP Support
==========================

> To build the sample to use CoAP instead of MQTT, use the ``-DOVERLAY_CONFIG=overlay_coap.conf`` options.

Reviewer: ```suggestion
To build the sample to use CoAP instead of MQTT, use the ``-DOVERLAY_CONFIG=overlay_coap.conf`` option.
```

   Modem responses longer than this length will be replaced with an error code message (-NRF_E2BIG).
   Cannot be less than 40 bytes.

When using CoAP, the following additional configuration options are available:

.. _CONFIG_COAP_SHADOW_CHECK_RATE_SECONDS:

> CONFIG_COAP_SHADOW_CHECK_RATE_SECONDS - Rate to check for shadow changes
>    How many seconds between requests for any change (delta) in the device shadow.

Reviewer 2: ```suggestion
CONFIG_COAP_SHADOW_CHECK_RATE_SECONDS - Rate of checking for shadow changes (seconds)
   Sets the time (in seconds) between requests for any change (delta) in the device shadow.
```

.. _CONFIG_COAP_SHADOW_THREAD_STACK_SIZE:

CONFIG_COAP_SHADOW_THREAD_STACK_SIZE - CoAP Shadow Thread Stack Size (bytes)

>    Sets the stack size (in bytes) for the shadow delta checking thread of the sample.

Reviewer 2: ```suggestion
   Sets the stack size (in bytes) for the thread that checks the shadow delta.
```

.. _CONFIG_NRF_CLOUD_COAP_FOTA:

> CONFIG_NRF_CLOUD_COAP_FOTA - Enable FOTA with CoAP
>    The sample periodically checks for pending FOTA jobs.
>    The sample performs the FOTA update when received.

Reviewer 2: ```suggestion
CONFIG_NRF_CLOUD_COAP_FOTA - FOTA with CoAP
   Enables periodical checks for pending FOTA jobs.
   The sample performs any received FOTA updates.
```

> If :kconfig:option:`CONFIG_NRF_CLOUD_COAP_FOTA` is enabled, these options additional are available:

Reviewer 2: ```suggestion
If :kconfig:option:`CONFIG_NRF_CLOUD_COAP_FOTA` is enabled, these additional options are available:
```

.. _CONFIG_COAP_FOTA_DL_TIMEOUT_MIN:

> CONFIG_COAP_FOTA_DL_TIMEOUT_MIN - CoAP FOTA download timeout
>     The time in minutes allotted for a FOTA download to complete.

Reviewer 2: ```suggestion
CONFIG_COAP_FOTA_DL_TIMEOUT_MIN - CoAP FOTA download timeout (minutes)
    Sets the time (in minutes) allotted for a FOTA download to complete.
```

.. _CONFIG_COAP_FOTA_USE_NRF_CLOUD_SETTINGS_AREA:

CONFIG_COAP_FOTA_USE_NRF_CLOUD_SETTINGS_AREA - Make FOTA compatible with other samples
   Use the same settings area as the nRF Cloud FOTA library.

.. _CONFIG_COAP_FOTA_SETTINGS_NAME:

CONFIG_COAP_FOTA_SETTINGS_NAME - Settings identifier

>    Set the identifier for the CoAP FOTA storage if :kconfig:option:`CONFIG_COAP_FOTA_USE_NRF_CLOUD_SETTINGS_AREA` is not enabled.

Reviewer 2: ```suggestion
   Sets the identifier for the CoAP FOTA storage if :kconfig:option:`CONFIG_COAP_FOTA_USE_NRF_CLOUD_SETTINGS_AREA` is not enabled.
```

.. _CONFIG_COAP_FOTA_SETTINGS_KEY_PENDING_JOB:

CONFIG_COAP_FOTA_SETTINGS_KEY_PENDING_JOB - Settings item key

>    Set the settings item key for pending FOTA job info if :kconfig:option:`CONFIG_COAP_FOTA_USE_NRF_CLOUD_SETTINGS_AREA` is not enabled.

Reviewer 2: ```suggestion
   Sets the settings item key for pending FOTA job info if :kconfig:option:`CONFIG_COAP_FOTA_USE_NRF_CLOUD_SETTINGS_AREA` is not enabled.
```

.. _CONFIG_COAP_FOTA_JOB_CHECK_RATE_MINUTES:

CONFIG_COAP_FOTA_JOB_CHECK_RATE_MINUTES - FOTA job check interval (minutes)

>    How many minutes between requests for a FOTA job.

Reviewer 2: ```suggestion
   Sets the time (in minutes) between requests for a FOTA job.
```

.. _CONFIG_COAP_FOTA_THREAD_STACK_SIZE:

CONFIG_COAP_FOTA_THREAD_STACK_SIZE - CoAP FOTA Thread Stack Size (bytes)

>    Sets the stack size (in bytes) for the FOTA job checking thread of the sample.

Reviewer 2: ```suggestion
   Sets the stack size (in bytes) for the thread that checks for a FOTA job.
```

.. include:: /libraries/modem/nrf_modem_lib/nrf_modem_lib_trace.rst
   :start-after: modem_lib_sending_traces_UART_start
   :end-before: modem_lib_sending_traces_UART_end

.. _nrf_cloud_multi_service_building_and_running:

Building and running
********************

.. |sample path| replace:: :file:`samples/cellular/nrf_cloud_multi_service`

.. include:: /includes/build_and_run_ns.txt

.. _nrf_cloud_multi_service_coap_provisioning:

Provisioning a CoAP device on nRF Cloud
***************************************

> For more information, see `nRF Cloud Utilities documentation`_.

Reviewer: This might have a better place at the end of this section.

Author: I specifically chose this location to emphasize that this provisioning must occur.  I do not want users to follow the "Building With CoAP Support" directions without properly provisioning the device.

Reviewer: Okay, but you might emphasize this a bit more, maybe saying something like: Read this first.

Author: I realized after viewing the built docs that the indent level of the Provisioning a CoAP device on nRF Cloud section was the same as the Building and running section.  I have swapped the two and it looks better to me now.

QA discovered a lack of documentation on a particular provisioning topic, so I added a 3 more sentences to the Provisioning a CoAP device on nRF Cloud.  If it needs changes, I can do a follow on PR.

First, create a self-signed CA certificate using the following command:

.. code-block:: console

   python3 create_ca_cert.py -c US -st OR -l Portland -o "My Company" -ou "RD" -cn example.com -e admin@example.com -f CA

Remember to set all the variables with your own information.

You should now have three files:

* :file:`<certificate_serial>_ca.pem`
* :file:`<certificate_serial>_prv.pem`
* :file:`<certificate_serial>_pub.pem`

Where ``<certificate_serial>`` is your certificate's serial number in hex.
You will use these files to sign device certificates.

You only need to generate these three files once.

> They can be used to sign as many device certificates as you need.

Reviewer: ```suggestion
You can use them to sign as many device certificates as you need.
```

Next, for each device you wish to provision, execute the following command in the same folder as the above command:

.. note::

>    The options used below assume that the sample is built with the default :kconfig:option:`CONFIG_NRF_CLOUD_CLIENT_ID_SRC_IMEI` option enabled and :kconfig:option:`CONFIG_NRF_CLOUD_CLIENT_ID_PREFIX` option set to ``nrf-``.

Reviewer: ```suggestion
   The options used in the example command assume that the sample is built with the default :kconfig:option:`CONFIG_NRF_CLOUD_CLIENT_ID_SRC_IMEI` option enabled and :kconfig:option:`CONFIG_NRF_CLOUD_CLIENT_ID_PREFIX` option set to ``nrf-``.
```

Deselect the :kconfig:option:`CONFIG_LOG_BACKEND_NRF_CLOUD_OUTPUT_TEXT` Kconfig option and select :kconfig:option:`CONFIG_LOG_BACKEND_NRF_CLOUD_OUTPUT_DICTIONARY` instead.
See `Dictionary-based Logging`_ to learn how dictionary-based logging works, how the dictionary is built, and how to decode the binary log output.

.. _nrf_cloud_multi_service_dependencies:

References
**********

> `RFC 7252 - The Constrained Application Protocol`_

Reviewer: Use bullet points.
